### ISSUE TITLE
Should we allow `shape.volume(G4String)` and `shape.place(G4string)`

### ISSUE THREAD

**Issue by gonzaponte** [jacg/nain4#94]
... as a proxy for `.volume(n4::material(G4String))` and `.place(n4::material(G4String))`?

`n4::volume(G4VSolid*, G4String)` could also be implemented, but `n4::place` doesn't have that possibility.

Pro: convenient
Cons:
+ may incur some inefficiencies as `n4::material` would be potentially called much more often than necessary.
+ increases the asymmetry between `n4::shape::place` and `n4::shape`

**Comment by jacg** [comment]
Hmm, convenient but ugly.

I don't see any major dangers. 

I doubt that the performance hit would be significant, but I guess we should measure it.

It would make n4 one-liners look even more convenient WRT G4 than they already do, but in real, non-trivial code I suspect it would be ugly.

Dunno, I'm ambivalent about it.

**Comment by gonzaponte** [comment]
> It would make n4 one-liners look even more convenient WRT G4 than they already do

This was my trigger. In the documentation looks very appealing to write a n4 one-liner instead of the corresponding G4 monstrosity.

**Comment by jacg** [comment]
Yes, and even though that would make n4 look even more convenient, it would teach people to do the wrong thing. So, if making n4 look good in one-liners is our main motivation for this, then I'd prefer not to do it.

**Comment by gonzaponte** [comment]
> it would teach people to do the wrong thing

what's the wrong thing?

I believe that real code can benefit from this syntax too. Most materials used in real scenarios are just looked for and plugged into a logical volume.

**Comment by jacg** [comment]
Using strings instead of symbols, to represent a finite number of values, is generally wrong. The compiler cannot help you if you misspell a string, but it can if you misspell a symbol. And you're comparing value rather than identity, so it's slower, too.

The material database should have a massive enum with all the allowed materials, rather than taking strings. OK, the fact that you can add new materials complicates this a bit, but I'm sure that with a bit of thought I could come up with a few symbol-based solutions which are much better than using strings to specify a finite number of values.

Because G4's API takes a string here, there's not much we can do about it, and passing a string if you want to use the material just once, is OK. But *as soon as* you want to use it more than once (in the same scope) then you really should define a local symbol bound to the material you want, and reuse that.

If our examples mostly use strings, the users will implicitly believe that that is the way to do it, and we'll be doing them a disservice.

> Most materials used in real scenarios are just looked for and plugged into a logical volume.

I suspect that in non-trivial geometries, they are plugged into multiple volumes. And then we really want symbols not strings.

**Comment by gonzaponte** [comment]
Agreed.

Should we rename this issue and keep this idea in the background?

**Comment by jacg** [comment]
Keep in the background: yes?

Rename: What's wrong with the current name? What would you call it instead?

**Comment by gonzaponte** [comment]
> What's wrong with the current name? What would you call it instead?

That, if we implement something, it would be with symbols instead of strings, which is what the title suggests.

**Comment by jacg** [comment]
Oh, I see, you actually want to implement a symbol interface to the G4 materials. No, I think that would be insane.

G4 should provide symbols in the first place. They don't. So *we* would have to track down all of their materials, and create the enum ourselves, and take on the burden of checking their lists of materials on *every single release* of G4, and update it. No thanks.

**Comment by gonzaponte** [comment]
I don't think it would be such a big deal. It could be automatized so not even the first time it would be a huge work. And I don't think new materials are introduced often enough for this to be a problem. I'm even thinking that this could be ignored until we get an issue (or preferably a PR) requesting new additions.  But ok, I take that you don't think it's worth the effort.

**Comment by jacg** [comment]
If we can find a reliable way of automating it, then it could be worthwhile.

Then we have to come up with a scheme for accessing user-defined materials in a way that doesn't break symmetry too much.
